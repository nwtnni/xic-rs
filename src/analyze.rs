mod anticipated_expressions;
mod available_expressions;
mod call_graph;
mod conditional_constant_propagation;
mod constant_propagation;
mod copy_propagation;
mod dot;
mod earliest;
mod latest;
mod live_ranges;
mod live_variables;
mod postponable_expressions;
mod used_expressions;
mod valid_allocation;

pub use anticipated_expressions::AnticipatedExpressions;
pub use available_expressions::AvailableExpressions;
pub use call_graph::CallGraph;
pub use conditional_constant_propagation::ConditionalConstantPropagation;
pub use conditional_constant_propagation::Constant;
pub use conditional_constant_propagation::Reachable;
pub use constant_propagation::ConstantPropagation;
pub use copy_propagation::CopyPropagation;
pub use dot::display;
pub use earliest::Earliest;
pub use latest::Latest;
pub use live_ranges::LiveRanges;
pub use live_ranges::Range;
pub use live_variables::LiveVariables;
pub use postponable_expressions::PostponableExpressions;
pub use used_expressions::UsedExpressions;

use std::collections::VecDeque;

use petgraph::Direction;

use crate::cfg;
use crate::cfg::Cfg;
use crate::cfg::Function;
use crate::data::operand::Label;
use crate::util;
use crate::Map;
use crate::Set;

pub trait Analysis<T: Function>: Sized {
    const BACKWARD: bool;

    type Data: Clone + Eq;

    fn default(&self) -> Self::Data;

    fn default_with_metadata(&self, _label: &Label) -> Self::Data {
        self.default()
    }

    fn transfer(&self, statement: &T::Statement, output: &mut Self::Data);

    fn transfer_with_metadata(
        &self,
        _label: &Label,
        _index: usize,
        statement: &T::Statement,
        output: &mut Self::Data,
    ) {
        self.transfer(statement, output);
    }

    fn merge<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a;

    fn merge_with_metadata<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = (&'a cfg::Edge, Option<&'a Self::Data>)>,
        Self::Data: 'a,
    {
        self.merge(outputs.map(|(_, data)| data), input)
    }
}

pub struct Solution<A: Analysis<T>, T: Function> {
    pub analysis: A,
    pub inputs: Map<Label, A::Data>,
    pub outputs: Map<Label, A::Data>,
}

pub fn analyze_default<A, T>(cfg: &Cfg<T>) -> Solution<A, T>
where
    A: Analysis<T> + Default,
    T: Function,
{
    analyze(Default::default(), cfg)
}

pub fn analyze<A: Analysis<T>, T: Function>(analysis: A, cfg: &Cfg<T>) -> Solution<A, T> {
    log::info!(
        "[{}] Running {} analysis on {}...",
        std::any::type_name::<Cfg<T>>(),
        std::any::type_name::<A>(),
        cfg.name(),
    );
    util::time!(
        "[{}] Done running {} analysis on {}",
        std::any::type_name::<Cfg<T>>(),
        std::any::type_name::<A>(),
        cfg.name(),
    );

    let strongly_connected_components = cfg.strongly_connected_components(A::BACKWARD);
    let (predecessors, successors) = match A::BACKWARD {
        true => (Direction::Outgoing, Direction::Incoming),
        false => (Direction::Incoming, Direction::Outgoing),
    };

    let mut inputs = Map::<Label, A::Data>::default();
    let mut outputs = Map::<Label, A::Data>::default();
    let mut worklist = VecDeque::new();
    let mut workset = Set::default();

    // Topological order
    for component in strongly_connected_components.into_iter().rev() {
        // Reverse postorder
        worklist.extend(component.iter().copied().rev());
        workset.extend(component.iter().copied().rev());

        let component = component.into_iter().collect::<Set<_>>();

        while let Some(label) = worklist.pop_front() {
            workset.remove(&label);

            let input = inputs
                .entry(label)
                .or_insert_with(|| analysis.default_with_metadata(&label));

            analysis.merge_with_metadata(
                cfg.edges_directed(predecessors, &label)
                    .map(|(predecessor, edge)| (edge, outputs.get(&predecessor))),
                input,
            );

            let mut output = input.clone();

            if A::BACKWARD {
                for (index, statement) in cfg[&label].iter().enumerate().rev() {
                    analysis.transfer_with_metadata(&label, index, statement, &mut output);
                }
            } else {
                for (index, statement) in cfg[&label].iter().enumerate() {
                    analysis.transfer_with_metadata(&label, index, statement, &mut output);
                }
            }

            match outputs.get_mut(&label) {
                Some(existing) if *existing == output => continue,
                Some(existing) => *existing = output,
                None => {
                    outputs.insert(label, output);
                }
            };

            cfg.neighbors(successors, &label)
                .filter(|successor| component.contains(successor))
                .filter(|successor| workset.insert(*successor))
                .for_each(|successor| worklist.push_back(successor));
        }
    }

    Solution {
        analysis,
        inputs,
        outputs,
    }
}
