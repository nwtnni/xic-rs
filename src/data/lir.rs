use std::fmt;

use crate::abi;
use crate::data::ir;
use crate::data::operand::Immediate;
use crate::data::operand::Label;
use crate::data::operand::Register;
use crate::data::operand::Temporary;
use crate::data::sexp;
use crate::data::sexp::Serialize;
use crate::data::symbol::Symbol;

pub const ZERO: Expression = Expression::Immediate(Immediate::Integer(0));
pub const ONE: Expression = Expression::Immediate(Immediate::Integer(1));
pub const EIGHT: Expression = Expression::Immediate(Immediate::Integer(abi::WORD));

pub type Unit<T> = ir::Unit<Function<T>>;

impl<T: Target> fmt::Display for Unit<T> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug)]
pub struct Function<T: Target> {
    pub name: Symbol,
    pub statements: Vec<Statement<T>>,
    pub arguments: usize,
    pub returns: usize,
    pub linkage: ir::Linkage,
    pub enter: T::Access,
    pub exit: T::Access,
}

impl<T: Target> Function<T> {
    pub fn callee_arguments(&self) -> Option<usize> {
        self.statements
            .iter()
            .filter_map(|statement| match statement {
                Statement::Call(_, arguments, _) => Some(arguments.len()),
                _ => None,
            })
            .max()
    }
}

impl<T: Target> fmt::Display for Function<T> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Expression {
    Immediate(Immediate),
    Temporary(Temporary),
    Memory(Box<Expression>),
    Binary(ir::Binary, Box<Expression>, Box<Expression>),
}

impl fmt::Display for Expression {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

impl From<i64> for Expression {
    fn from(integer: i64) -> Self {
        Expression::Immediate(Immediate::Integer(integer))
    }
}

impl From<Label> for Expression {
    fn from(label: Label) -> Self {
        Expression::Immediate(Immediate::Label(label))
    }
}

impl From<Temporary> for Expression {
    fn from(temporary: Temporary) -> Self {
        Expression::Temporary(temporary)
    }
}

impl From<Register> for Expression {
    fn from(register: Register) -> Self {
        Expression::Temporary(Temporary::Register(register))
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Statement<T> {
    Jump(Label),
    CJump {
        condition: ir::Condition,
        left: Expression,
        right: Expression,
        r#true: Label,
        r#false: T,
    },
    Call(Expression, Vec<Expression>, usize),
    Label(Label),
    Move {
        destination: Expression,
        source: Expression,
    },
    Return(Vec<Expression>),
}

impl<T: Serialize> fmt::Display for Statement<T> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt, "{}", self.sexp())
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Fallthrough;

pub trait Target: sexp::Serialize + Copy + Clone + Eq + fmt::Debug {
    type Access;
    fn target(&self) -> Option<&Label>;
    fn target_mut(&mut self) -> Option<&mut Label>;
    fn access(access: &Self::Access) -> Option<&Label>;
}

impl Target for Fallthrough {
    type Access = Label;

    fn target(&self) -> Option<&Label> {
        None
    }

    fn target_mut(&mut self) -> Option<&mut Label> {
        None
    }

    fn access(access: &Self::Access) -> Option<&Label> {
        Some(access)
    }
}

impl Target for Label {
    type Access = ();

    fn target(&self) -> Option<&Label> {
        Some(self)
    }

    fn target_mut(&mut self) -> Option<&mut Label> {
        Some(self)
    }

    fn access((): &Self::Access) -> Option<&Label> {
        None
    }
}

#[macro_export]
macro_rules! lir {
    ((CONST $($integer:tt)+)) => {
        $crate::data::lir::Expression::from(
            $crate::data::lir::lir!($($integer)+)
        )
    };
    ((NAME $($label:tt)+)) => {
        $crate::data::lir::Expression::from(
            $crate::data::lir::lir!($($label)+)
        )
    };
    ((TEMP $($temporary:tt)+)) => {
        $crate::data::lir::Expression::from(
            $crate::data::lir::lir!($($temporary)+)
        )
    };
    ((MEM $expression:tt)) => {
        $crate::data::lir::Expression::Memory(Box::new(
            $crate::data::lir::lir!($expression)
        ))
    };

    ((JUMP $label:ident)) => {
        $crate::data::lir::Statement::Jump($label)
    };
    ((CJUMP ($condition:ident $left:tt $right:tt) $r#true:ident)) => {
        $crate::data::lir::Statement::CJump {
            condition: $crate::data::ir::ir!($condition),
            left: $crate::data::lir::lir!($left),
            right: $crate::data::lir::lir!($right),
            r#true: $r#true,
            r#false: $crate::data::lir::Fallthrough,
        }
    };
    ((CJUMP ($condition:ident $left:tt $right:tt) $r#true:ident $r#false:ident)) => {
        $crate::data::lir::Statement::CJump {
            condition: $crate::data::ir::ir!($condition),
            left: $crate::data::lir::lir!($left),
            right: $crate::data::lir::lir!($right),
            r#true: $r#true,
            r#false: $r#false,
        }
    };
    ((CALL $function:tt $returns:tt $($argument:tt)*)) => {
        $crate::data::lir::Statement::Call(
            $crate::data::lir::lir!($function),
            vec![
                $(
                    $crate::data::lir::lir!($argument),
                )*
            ],
            $returns,
        )
    };
    ((LABEL $label:tt)) => {
        $crate::data::lir::Statement::Label($label)
    };
    ((MOVE $into:tt $from:tt)) => {
        $crate::data::lir::Statement::Move {
            destination: $crate::data::lir::lir!($into),
            source: $crate::data::lir::lir!($from),
        }
    };
    ((RETURN $($expression:tt)*)) => {
        $crate::data::lir::Statement::Return(
            vec![
                $(
                    $crate::data::lir::lir!($expression),
                )*
            ]
        )
    };
    (($binary:ident $left:tt $right:tt)) => {
        $crate::data::lir::Expression::Binary(
            $crate::data::ir::ir!($binary),
            Box::new($crate::data::lir::lir!($left)),
            Box::new($crate::data::lir::lir!($right)),
        )
    };

    ($expression:expr) => {
        $expression
    }
}

// https://github.com/rust-lang/rust/pull/52234#issuecomment-976702997
#[doc(hidden)]
pub use lir;
