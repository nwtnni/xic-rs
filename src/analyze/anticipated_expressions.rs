use std::collections::BTreeSet;

use crate::analyze::Analysis;
use crate::data::lir;

pub struct AnticipatedExpressions;

impl<T: lir::Target> Analysis<lir::Function<T>> for AnticipatedExpressions {
    const BACKWARD: bool = true;

    type Data = BTreeSet<lir::Expression>;

    fn new() -> Self {
        Self
    }

    fn default(&self) -> Self::Data {
        BTreeSet::new()
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match statement {
            lir::Statement::Jump(_) | lir::Statement::Label(_) => (),
            lir::Statement::CJump {
                condition: _,
                left,
                right,
                r#true: _,
                r#false: _,
            } => {
                output.insert(left.clone());
                output.insert(right.clone());
            }
            lir::Statement::Call(function, arguments, returns) => {
                for r#return in 0..*returns {
                    Self::remove(output, &lir::Expression::Return(r#return));
                }

                output.extend(arguments.iter().cloned());
                output.insert(function.clone());
            }
            lir::Statement::Move {
                destination,
                source,
            } => {
                Self::remove(output, destination);
                output.insert(source.clone());
            }
            lir::Statement::Return(returns) => {
                output.extend(returns.iter().cloned());
            }
        }
    }

    fn merge<'a, I>(&self, mut outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        input.clear();
        input.extend(outputs.next().into_iter().flatten().flatten().cloned());

        let mut stack = Vec::new();

        for output in outputs.flatten() {
            stack.extend(
                input
                    .iter()
                    .filter(|expression| !output.contains(expression))
                    // Could be avoided with https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.drain_filter
                    .cloned(),
            );

            stack.drain(..).for_each(|kill| Self::remove(input, &kill));
        }
    }
}

impl AnticipatedExpressions {
    pub(super) fn remove(output: &mut BTreeSet<lir::Expression>, kill: &lir::Expression) {
        output.remove(kill);

        let mut stack = vec![kill.clone()];

        while let Some(killed) = stack.pop() {
            output.retain(|kill| {
                if !Self::contains(kill, &killed) {
                    return true;
                }

                stack.push(kill.clone());
                false
            })
        }
    }

    fn contains(expression: &lir::Expression, killed: &lir::Expression) -> bool {
        match expression {
            lir::Expression::Argument(_)
            | lir::Expression::Return(_)
            | lir::Expression::Immediate(_)
            | lir::Expression::Temporary(_) => expression == killed,
            lir::Expression::Memory(address) => Self::contains(&*address, killed),
            lir::Expression::Binary(_, left, right) => {
                Self::contains(&*left, killed) || Self::contains(&*right, killed)
            }
        }
    }
}
