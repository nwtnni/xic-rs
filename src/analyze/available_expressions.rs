use std::collections::BTreeSet;

use crate::analyze::analyze;
use crate::analyze::Analysis;
use crate::analyze::AnticipatedExpressions;
use crate::analyze::Solution;
use crate::data::lir;
use crate::data::operand::Label;

pub struct AvailableExpressions<T: lir::Target> {
    anticipated: Solution<AnticipatedExpressions, lir::Function<T>>,
}

impl<T: lir::Target> Analysis<lir::Function<T>> for AvailableExpressions<T> {
    const BACKWARD: bool = false;

    type Data = BTreeSet<lir::Expression>;

    fn new() -> Self {
        unreachable!()
    }

    fn new_with_metadata(cfg: &crate::cfg::Cfg<lir::Function<T>>) -> Self {
        Self {
            anticipated: analyze(cfg),
        }
    }

    fn default(&self) -> Self::Data {
        unreachable!()
    }

    fn default_with_metadata(&self, label: &Label) -> Self::Data {
        self.anticipated.outputs[label].clone()
    }

    fn transfer(&self, statement: &lir::Statement<T>, output: &mut Self::Data) {
        match statement {
            lir::Statement::Jump(_)
            | lir::Statement::CJump { .. }
            | lir::Statement::Label(_)
            | lir::Statement::Return(_) => (),
            lir::Statement::Call(_, _, returns) => {
                for r#return in 0..*returns {
                    AnticipatedExpressions::remove(output, &lir::Expression::Return(r#return));
                }
            }
            lir::Statement::Move {
                destination,
                source: _,
            } => {
                AnticipatedExpressions::remove(output, destination);
            }
        }
    }

    fn merge<'a, I>(&self, outputs: I, input: &mut Self::Data)
    where
        I: Iterator<Item = Option<&'a Self::Data>>,
        Self::Data: 'a,
    {
        <AnticipatedExpressions as Analysis<lir::Function<T>>>::merge(
            &self.anticipated.analysis,
            outputs,
            input,
        );
    }
}
