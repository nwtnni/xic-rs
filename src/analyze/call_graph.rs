use petgraph::graphmap::DiGraphMap;
use petgraph::visit::Walker as _;
use petgraph::Direction;

use crate::data::lir;
use crate::data::operand::{Immediate, Label};
use crate::data::symbol::Symbol;

pub struct CallGraph(DiGraphMap<Symbol, ()>);

impl CallGraph {
    pub fn new<T: lir::Target>(lir: &lir::Unit<T>) -> Self {
        let mut graph = lir
            .functions
            .iter()
            .flat_map(|(caller, function)| {
                function
                    .statements
                    .iter()
                    .map(move |statement| (caller, statement))
            })
            .filter_map(|(caller, statement)| match statement {
                lir::Statement::Call(
                    lir::Expression::Immediate(Immediate::Label(Label::Fixed(callee))),
                    _,
                    _,
                ) => Some((*caller, *callee)),
                _ => None,
            })
            .collect::<DiGraphMap<_, _>>();

        for name in lir.functions.keys() {
            graph.add_node(*name);
        }

        Self(graph)
    }

    pub fn is_recursive(&self, name: &Symbol) -> bool {
        self.0.contains_edge(*name, *name)
    }

    pub fn is_leaf(&self, name: &Symbol) -> bool {
        self.0
            .neighbors_directed(*name, Direction::Outgoing)
            .count()
            == 0
    }

    pub fn postorder(&self, start: &Symbol) -> impl Iterator<Item = Symbol> + '_ {
        petgraph::visit::DfsPostOrder::new(&self.0, *start).iter(&self.0)
    }
}
