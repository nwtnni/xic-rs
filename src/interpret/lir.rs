use std::io;

use anyhow::anyhow;
use anyhow::Context as _;

use crate::constants;
use crate::data::ir;
use crate::data::lir;
use crate::data::operand;
use crate::data::sexp::Serialize as _;
use crate::data::symbol;
use crate::interpret::postorder;
use crate::interpret::postorder::Postorder;
use crate::interpret::Global;
use crate::interpret::Local;
use crate::interpret::Operand;
use crate::interpret::Value;

pub fn interpret_lir<'io, R, W, T>(
    unit: &ir::Unit<lir::Function<T>>,
    stdin: R,
    stdout: W,
) -> anyhow::Result<()>
where
    R: io::BufRead + 'io,
    W: io::Write + 'io,
    T: lir::Target,
{
    let unit = Postorder::traverse_lir_unit(unit);

    let mut global = Global::new(&unit.data, stdin, stdout);
    let mut local = Local::new(
        &unit,
        &symbol::intern_static(constants::XI_MAIN),
        &[Value::Integer(0)],
    );

    assert!(local.interpret_lir(&unit, &mut global)?.is_empty());

    Ok(())
}

impl<'a, T: lir::Target> Local<'a, postorder::Lir<'a, T>> {
    fn interpret_lir(
        &mut self,
        unit: &ir::Unit<Postorder<postorder::Lir<'a, T>>>,
        global: &mut Global,
    ) -> anyhow::Result<Vec<Value>> {
        loop {
            let instruction = match self.step() {
                Some(instruction) => instruction,
                None => return Ok(Vec::new()),
            };

            match instruction {
                postorder::Lir::Expression(expression) => {
                    self.interpret_expression(global, expression)?
                }
                postorder::Lir::Statement(statement) => {
                    if let Some(returns) = self.interpret_statement(unit, global, statement)? {
                        return Ok(returns);
                    }
                }
            }
        }
    }

    fn interpret_expression(
        &mut self,
        global: &mut Global,
        expression: &lir::Expression,
    ) -> anyhow::Result<()> {
        log::trace!("E> {}", expression.sexp());
        match expression {
            lir::Expression::Immediate(operand::Immediate::Integer(integer)) => {
                self.push(Operand::Integer(*integer))
            }
            lir::Expression::Immediate(operand::Immediate::Label(label)) => {
                self.push(Operand::Label(*label, 8))
            }
            lir::Expression::Temporary(temporary) => self.push(Operand::Temporary(*temporary)),
            lir::Expression::Argument(index) => {
                let argument = self.get_argument(*index).into_operand();
                self.push(argument);
            }
            lir::Expression::Return(index) => {
                let r#return = self.get_return(*index).into_operand();
                self.push(r#return);
            }
            lir::Expression::Memory(_) => {
                let address = self.pop(global);
                self.push(Operand::Memory(address));
            }
            lir::Expression::Binary(binary, _, _) => self.interpret_binary(global, binary),
        }

        Ok(())
    }

    fn interpret_statement(
        &mut self,
        unit: &ir::Unit<Postorder<postorder::Lir<'a, T>>>,
        global: &mut Global,
        statement: &lir::Statement<T>,
    ) -> anyhow::Result<Option<Vec<Value>>> {
        log::debug!("S> {}", statement.sexp());
        match statement {
            lir::Statement::Label(_) => unreachable!(),
            lir::Statement::Jump(label) => {
                self.interpret_jump(label);
            }
            lir::Statement::CJump {
                condition,
                left: _,
                right: _,
                r#true,
                r#false,
            } => {
                if self.interpret_condition(global, condition) {
                    self.interpret_jump(r#true);
                } else if let Some(label) = r#false.label() {
                    self.interpret_jump(label);
                }
            }
            lir::Statement::Call(_, arguments, _) => {
                let arguments = self.pop_list(global, arguments.len());
                let name = self.pop_name(global);

                log::info!("Calling function {} with arguments {:?}", name, arguments);

                let returns = global
                    .interpret_library(name, &arguments)
                    .unwrap_or_else(|| {
                        Local::new(unit, &name, &arguments).interpret_lir(unit, global)
                    })
                    .with_context(|| anyhow!("Calling function {}", name))?;

                if let Some(r#return) = returns.first() {
                    self.push(r#return.into_operand());
                }

                self.insert_returns(&returns);
            }
            lir::Statement::Move {
                destination: _,
                source: _,
            } => self.interpret_move(global),
            lir::Statement::Return(returns) => {
                return Ok(Some(self.pop_list(global, returns.len())));
            }
        }

        Ok(None)
    }
}
