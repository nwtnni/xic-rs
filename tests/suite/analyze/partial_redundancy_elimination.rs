use xic::api::analyze::analyze;
use xic::api::analyze::display;
use xic::api::analyze::AnticipatedExpressions;
use xic::api::analyze::AvailableExpressions;
use xic::api::analyze::Earliest;
use xic::api::analyze::Latest;
use xic::api::analyze::PostponableExpressions;
use xic::api::analyze::UsedExpressions;
use xic::data::lir::Fallthrough;
use xic::data::lir::Function;

macro_rules! partial_redundancy_elimination {
    ($function:ident $($tt:tt)*) => {
        #[test]
        fn $function() {
            let function = lir_function!($function $($tt)*);
            let (anticipated, available, earliest, postponable, latest, used) = partial_redundancy_elimination(function);
            insta::assert_display_snapshot!(concat!(stringify!($function), "_anticipated"), anticipated);
            insta::assert_display_snapshot!(concat!(stringify!($function), "_available"), available);
            insta::assert_display_snapshot!(concat!(stringify!($function), "_earliest"), earliest);
            insta::assert_display_snapshot!(concat!(stringify!($function), "_postponable"), postponable);
            insta::assert_display_snapshot!(concat!(stringify!($function), "_latest"), latest);
            insta::assert_display_snapshot!(concat!(stringify!($function), "_used"), used);
        }
    }
}

fn partial_redundancy_elimination(
    function: Function<Fallthrough>,
) -> (String, String, String, String, String, String) {
    let cfg = xic::api::construct_cfg(function);

    let solution = analyze::<AnticipatedExpressions, _>(&cfg);
    let anticipated = super::super::graph(display(&solution, &cfg));

    let solution = analyze::<AvailableExpressions<_>, _>(&cfg);
    let available = super::super::graph(display(&solution, &cfg));

    let solution = analyze::<Earliest<_>, _>(&cfg);
    let earliest = super::super::graph(display(&solution, &cfg));

    let solution = analyze::<PostponableExpressions<_>, _>(&cfg);
    let postponable = super::super::graph(display(&solution, &cfg));

    let solution = analyze::<Latest<_>, _>(&cfg);
    let latest = super::super::graph(display(&solution, &cfg));

    let solution = analyze::<UsedExpressions<_>, _>(&cfg);
    let used = super::super::graph(display(&solution, &cfg));

    (anticipated, available, earliest, postponable, latest, used)
}

partial_redundancy_elimination! {
    used_on_one_branch: 0 -> 0;
    temporaries: a, b, c;
    labels: branch, fallthrough, merge, exit;
        (MOVE (TEMP b) (CONST 0))
        (CJUMP (EQ (CONST 0) (CONST 0)) branch)
    (LABEL fallthrough)
        (MOVE (TEMP a) (ADD (TEMP b) (CONST 1)))
        (JUMP merge)
    (LABEL branch)
        (MOVE (TEMP a) (CONST 0))
        (JUMP merge)
    (LABEL merge)
        (MOVE (TEMP c) (ADD (TEMP b) (CONST 1)))
        (JUMP exit)
}

partial_redundancy_elimination! {
    basic_loop: 0 -> 0;
    temporaries: a, b, c;
    labels: r#loop, split, fallthrough, exit;
        (MOVE (TEMP a) (CONST 0))
        (JUMP r#loop)
    (LABEL r#loop)
        (MOVE (TEMP c) (ADD (TEMP a) (TEMP b)))
        (CJUMP (EQ (CONST 0) (CONST 0)) split)
    (LABEL fallthrough)
        (JUMP exit)
    (LABEL split)
        (JUMP r#loop)
}

partial_redundancy_elimination! {
    // https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.92.4197&rep=rep1&type=pdf
    //
    // Note: loop between _12 and _13 omitted, as there's nothing interesting there.
    knoop_ruthing_steffen: 0 -> 0;
    temporaries: a, b, c, x, y;
    labels: _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _14, _15, _16, _17, _18, exit;
    (LABEL _1)
        (CJUMP (EQ (CONST 0) (CONST 0)) _2)
    (LABEL _4)
        (JUMP _5)
    (LABEL _2)
        (MOVE (TEMP a) (TEMP c))
        (JUMP _3)
    (LABEL _3)
        (MOVE (TEMP x) (ADD (TEMP a) (TEMP b)))
        (JUMP _5)
    (LABEL _5)
        (CJUMP (EQ (CONST 0) (CONST 0)) _6)
    (LABEL _7)
        (JUMP _18)
    (LABEL _6)
        (CJUMP (EQ (CONST 0) (CONST 0)) _8)

    (LABEL _9)
        (JUMP _12)
    (LABEL _12)
        (CJUMP (EQ (CONST 0) (CONST 0)) _17)
    (LABEL _15)
        (MOVE (TEMP y) (ADD (TEMP a) (TEMP b)))
        (JUMP _16)
    (LABEL _17)
        (MOVE (TEMP x) (ADD (TEMP a) (TEMP b)))
        (JUMP _18)

    (LABEL _8)
        (JUMP _11)
    (LABEL _11)
        (CJUMP (EQ (CONST 0) (CONST 0)) _14)
    (LABEL _10)
        (MOVE (TEMP y) (ADD (TEMP a) (TEMP b)))
        (JUMP _11)
    (LABEL _14)
        (JUMP _16)

    (LABEL _16)
        (MOVE (TEMP x) (ADD (TEMP a) (TEMP b)))
        (JUMP _18)
    (LABEL _18)
        (JUMP exit)
}

partial_redundancy_elimination! {
    call_argument: 0 -> 0;
    labels: black_box;
    (CALL (NAME black_box) 0 (ADD (CONST 1) (CONST 2)))
}

partial_redundancy_elimination! {
    induction_variable: 0 -> 0;
    temporaries: a;
    labels: r#loop;
        (MOVE (TEMP a) (CONST 0))
        (JUMP r#loop)
    (LABEL r#loop)
        (MOVE (TEMP a) (ADD (TEMP a) (CONST 1)))
        (CJUMP (GE (TEMP a) (CONST 5)) r#loop)
}

partial_redundancy_elimination! {
    induction_variable_regression: 0 -> 0;
    temporaries: a, b, c;
    labels: r#while, r#true, r#false, black_box, exit;
        (MOVE (TEMP a) (_ARG 0))
        (MOVE (TEMP b) (CONST 0))
        (JUMP r#while)
    (LABEL r#while)
        (CJUMP (GE (TEMP b) (CONST 3)) r#true)
    (LABEL r#false)
        (CALL (NAME black_box) 1 (TEMP b))
        (MOVE (TEMP c) (_RET 0))
        (CALL (NAME black_box) 0 (TEMP c))
        (MOVE (TEMP b) (ADD (TEMP b) (CONST 1)))
        (JUMP r#while)
    (LABEL r#true)
        (RETURN)
        (JUMP exit)
}
