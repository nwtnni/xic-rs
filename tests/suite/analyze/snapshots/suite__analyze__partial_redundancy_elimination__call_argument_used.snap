---
source: tests/suite/analyze/partial_redundancy_elimination.rs
assertion_line: 138
expression: used
---

                      call_argument

+---------------------------------------------------------+
|                         enter:                          |
| {}                                                      |
| (EXP (CALL (NAME black_box) (ADD (CONST 1) (CONST 2)))) |
| {}                                                      |
| (JUMP exit)                                             |
| {}                                                      |
+---------------------------------------------------------+
  |
  |
  v
+---------------------------------------------------------+
|                          exit:                          |
| {}                                                      |
+---------------------------------------------------------+

